### PR TITLE
ENG-12985: Delta tables used for joins in materialized views should never generate DR records

### PR DESCRIPTION
**Pull request by slowsoul**
The delta table used in joins in materialized views used to inherit the "DR" catalog attribute from its source table. If the source table is a DR table, so will the delta table, which essentially records all operations on delta table into the DR stream, which is undesirable because 1) it breaks DR binary log structural constraint in case of snapshot restore and 2) it causes the consumer cluster to diverge from the producer cluster.

To fix this, delta tables should always be set as non-DR tables.

### CODE REVIEW

File: src/ee/storage/TableCatalogDelegate.cpp [hunk 1]
```diff
@@ -73,12 +73,13 @@ Table* TableCatalogDelegate::getTable() const {
...
 TupleSchema* TableCatalogDelegate::createTupleSchema(catalog::Table const& catalogTable,
-                                                     bool isXDCR) {
+                                                     bool isXDCR,
+                                                     bool forceNoDR) {
     // Columns:
     // Column is stored as map<std::string, Column*> in Catalog. We have to
     // sort it by Column index to preserve column order.
     auto numColumns = catalogTable.columns().size();
-    bool needsDRTimestamp = isXDCR && catalogTable.isDRed();
+    bool needsDRTimestamp = isXDCR && !forceNoDR && catalogTable.isDRed();
```
Comment (yzhang1991) [inline src/ee/storage/TableCatalogDelegate.cpp]: When `createDeltaTable()` called `createTupleSchema()`, the `isXDCR` flag is set to false:
```c++
Table* deltaTable = constructTableFromCatalog(catalogDatabase, catalogTable, false, 1024 * 64);
```
The `needsDRTimestamp` is always false for the delta table case.
So `forceNoDR` is not needed here, I think you can remove it from this function.

File: src/ee/storage/TableCatalogDelegate.cpp [hunk 2]
```diff
@@ -279,7 +281,7 @@ Table* TableCatalogDelegate::constructTableFromCatalog(catalog::Database const&
     }
 
     // get the schema for the table
-    TupleSchema* schema = createTupleSchema(catalogTable, isXDCR);
+    TupleSchema* schema = createTupleSchema(catalogTable, isXDCR, forceNoDR);
```
Comment (yzhang1991) [inline src/ee/storage/TableCatalogDelegate.cpp]: Remove the `forceNoDR` flag here.

File: src/ee/storage/TableCatalogDelegate.hpp [hunk 3]
```diff
@@ -144,6 +145,7 @@ class TableCatalogDelegate {
     Table *constructTableFromCatalog(catalog::Database const &catalogDatabase,
                                      catalog::Table const &catalogTable,
                                      bool isXDCR,
+                                     bool forceNoDR = false,
```
Comment (yzhang1991) [inline src/ee/storage/TableCatalogDelegate.hpp]: Remove forceNoDR flag here

File: src/ee/storage/TableCatalogDelegate.cpp [hunk 4]
```diff
@@ -369,7 +371,7 @@ Table* TableCatalogDelegate::constructTableFromCatalog(catalog::Database const&
 
     bool exportEnabled = isExportEnabledForTable(catalogDatabase, tableId);
     bool tableIsExportOnly = isTableExportOnly(catalogDatabase, tableId);
-    bool drEnabled = catalogTable.isDRed();
+    bool drEnabled = !forceNoDR && catalogTable.isDRed();
```
Comment (yzhang1991) [inline src/ee/storage/TableCatalogDelegate.cpp]: This is the only place that we need the `forceNoDR` flag.

File: src/ee/storage/TableCatalogDelegate.cpp (hunk 1 above)
Comment (slowsoul) [inline src/ee/storage/TableCatalogDelegate.cpp]: I see, I will remove it.

File: src/ee/storage/TableCatalogDelegate.hpp (hunk 3 above)
Comment (slowsoul) [inline src/ee/storage/TableCatalogDelegate.hpp]: I think `forceNoDR` is still needed in `constructTableFromCatalog()`, right? Did you mean to add this comment on the declaration of `createTupleSchema()`

File: src/ee/storage/TableCatalogDelegate.hpp [hunk 5]
```diff
@@ -144,7 +144,8 @@ class TableCatalogDelegate {
     Table *constructTableFromCatalog(catalog::Database const &catalogDatabase,
```
Comment (yzhang1991) [inline src/ee/storage/TableCatalogDelegate.hpp]: Please add a comment here about what `forceNoDR` is and where we use it (Now only the delta table creation will use it).

File: src/ee/storage/TableCatalogDelegate.cpp [hunk 6]
```diff
@@ -446,7 +447,7 @@ PersistentTable* TableCatalogDelegate::createDeltaTable(catalog::Database const&
     // Delta table will only have one row (currently).
     // Set the table block size to 64KB to achieve better space efficiency.
     // FYI: maximum column count = 1024, largest fixed length data type is short varchars (64 bytes)
-    Table* deltaTable = constructTableFromCatalog(catalogDatabase, catalogTable, false, 1024 * 64);
```
Comment (yzhang1991) [inline src/ee/storage/TableCatalogDelegate.cpp]: Add a comment here that we will force the construction of the table to have no DR.

**Pull request merged**